File: 1715-split-a-string-into-the-max-number-of-unique-substrings/1715-split-a-string-into-the-max-number-of-unique-substrings.py
class Solution:
    def all_possible_substrings(self, s, pos, ms, us):
        if pos==len(s):
            ms[0]=max(ms[0],len(us))
            return 
        tempstr=""
        for i in range(pos,len(s)):
            tempstr+=s[i]
            if tempstr not in us:
                us.add(tempstr)
                self.all_possible_substrings(s, i+1, ms, us)
                us.remove(tempstr)

    
    def maxUniqueSplit(self, s: str) -> int:
        ms =[0]
        us = set()
        self.all_possible_substrings(s, 0, ms, us)
        return ms[0]

File: 1715-split-a-string-into-the-max-number-of-unique-substrings/1715-split-a-string-into-the-max-number-of-unique-substrings.cpp
class Solution {
    void allPossibleSubstrings(string& s, int pos, int& ms, unordered_set<string>& us) {
        if (pos == s.size()) {
            if (ms < us.size())
                ms = us.size();
            return;
        }  
        string temp;
        for (int i = pos; i < s.size(); ++i) {
            temp.push_back(s[i]);
            if (us.count(temp) == 0) {
                us.insert(temp);
                allPossibleSubstrings(s, i + 1, ms, us);
                us.erase(temp);
            }
        }
    }
    
public:
    int maxUniqueSplit(string s) {
        int ms = 0;
        unordered_set<string> us;
        allPossibleSubstrings(s, 0, ms, us);
        return ms;
    }
};
